### DISCUSSION TITLE
New user of phase field : How to model multiple grains ?

### DISCUSSION THREAD

**Discussion by AlexSacMorane** [idaholab/moose#18838] [Q&A Modules: Phase field] [answered]
Dear all,

I am a totally new user of MOOSE and with my supervisors we would like to use this software for our research. The main goal is to simulate a grains sample which is sheared (2D will be ok for a first step). We are interested in the phase field model implanted in Moose and we are trying to link it with a discrete element modelisation. I have a bit looked into test and example files from Moose but I have some questions :

- As our goal is to have one unconserved variable c by grain, is it possible to automate the input file ? I mean to not write 500 [./c] [../] sections in the Variables part for example if we have 500 grains. 
- Is it possible to detect contacts between grains and avoid an overlap between them? 
- How can we apply contact law between grains ? I have seen the function ComputeExternalGrainForceAndTorque but as it is undocumented I do not understand well how it is working
- I have tried to play with the function MultiGrainRigidBodyMotion (modules/phase_field/examples/rigidbodymotion/grain_motion_GT.i) but particles seem to disappear. As it is also undocumented I do not inderstand well how it is working

If you have some example files concerning those subjects, I will enjoy to read them.

Regards
Alexandre Sac--Morane

**Comment by laagesen** [comment, marked as answer]
Hi Alexandre- welcome to the MOOSE framework!

MOOSE has a system called Actions which can be used to automate adding other objects to input files. There are multiple actions that have been created to help with grain growth problems. Probably the best thing to do is to look at some example files where they are used. There are several in the directory `moose/modules/phase_field/examples/grain_growth/`. `grain_growth_2D_graintracker.i` would be a good one to start with. In this example 100 grains are added but the Action syntax allows for a compact input file.

We have a system for representing multiple grains with a reduced number of order parameter called GrainTracker. That system makes sure that grains represented by the same order parameter do not come into contact with each other, and remaps grains to different OP's to prevent that from happening. It's used in the previous example I listed. You can also read documentation here:

https://mooseframework.inl.gov/moose/source/postprocessors/GrainTracker.html

I'll see if I can ask others to help on the other 2 questions, as I'm not as familiar with the other objects you mentioned offhand...

**Reply by AlexSacMorane** [reply]
Hi, 
thanks for your answer. I am checking what you told me and I will let you know if I have other questions about it.

**Comment by SudiptaBiswas** [comment]
Hi Alex, 

The rigid body motion system allows you to apply forces and torques on particles for a given forces density at the interface. We don't have a direct way to account for contact forces on particles in phase field. You can find details about the rigid-body motion system here, https://link.springer.com/article/10.1007/s10853-017-1846-3

**Reply by AlexSacMorane** [reply]
Thanks for your answer. I will read this link.
